Update find UI when ImageAnalysisQueue empties

danae404 opened the pull request:
#### ce5226e9d21450f6ee5ed21e8969f81bc4dccdb9
<pre>
Update find UI when ImageAnalysisQueue empties
<a href="https://bugs.webkit.org/show_bug.cgi?id=243071">https://bugs.webkit.org/show_bug.cgi?id=243071</a>
rdar://97396153

Reviewed by Wenson Hsieh.

Test:
Updated FindInPageAPI.FindAPITextInImage to check that the first
Find-in-page operation on a webpage occurs twice.

This patch improves the internal setting ImageAnalysisForFindInPage.
Now the find ui is updated once the ImageAnalysisQueue is empty, so
matches from images show up shortly after other text matches without
the need for user interaction to update the ui. This is accomplished
by re-doing the most recent find-in-page when the image queue is newly
empty.

* Source/WebCore/PAL/pal/HysteresisActivity.h:

Added WTF_MAKE_FAST_ALLOCATED.

* Source/WebCore/page/ImageAnalysisQueue.cpp:
(WebCore::ImageAnalysisQueue::enqueueAllImages): Deleted.
(WebCore::ImageAnalysisQueue::enqueueAllImagesIfNeeded): Added.

Changed name of function and added m_analysisOfAllImagesOnPageHasStarted
Check.

(WebCore::ImageAnalysisQueue::resumeProcessing):

Now checks if the image queue is empty, and if so, triggers
m_imageQueueEmptyHysteresis.

(WebCore::ImageAnalysisQueue::setDidBecomeEmptyCallback): Added.

Initializes m_imageQueueEmptyHysteresis with given callback to be called
When the HysteresisActivity goes off.

(WebCore::ImageAnalysisQueue::clearDidBecomeEmptyCallback): Added.

(WebCore::ImageAnalysisQueue::clear):

Reset m_analysisOfAllImagesOnPageHasStarted and m_imageQueueEmptyHysteresis.

Sets m_imageQueueEmptyHysteresis to nullptr.

* Source/WebCore/page/ImageAnalysisQueue.h:

* Source/WebCore/page/Page.cpp:
(WebCore::Page::analyzeImagesForFindInPage):

If setting is enabled, calls ImageAnalysisQueue::setImageQueueEmptyHysteresis
with given callback, then enqueues all images on page
to be analyzed if they have not begun to be analyzed yet.

* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::findString):

Calls page::analyzeImagesForFindInPagee with a
callback function that calls FindController::findString with
Arguments configured to re-find the current string. Therefore there is a second
search for matches on the page and a UI update after the text from the images
has been added to the page.

findString has also been changed so that its completionHandler argument
Has a default value of { } and to take an enum shouldTriggerImageAnalysis.

(WebKit::FindController::hideFindUI):

Now also calls ImageAnalysisQueue::endImageQueueEmptyHysteresis.

* Source/WebKit/WebProcess/WebPage/FindController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp
(WebKit::WebPage::findString):

Now passes TriggerImageAnalysis::Yes to FindController::findString.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
(-[FindDelegate didFindMatches:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/image-with-text.html: Added.

Canonical link: <a href="https://commits.webkit.org/253244@main">https://commits.webkit.org/253244@main</a>
</pre>

cdumez commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -158,10 +159,19 @@ void ImageAnalysisQueue::resumeProcessing()
     }
 }
 
+void ImageAnalysisQueue::setImageQueueEmptyHysteresis(std::unique_ptr<PAL::HysteresisActivity>& timer)
```
Should use `std::unique_ptr<PAL::HysteresisActivity>&&`

cdumez commented on Source/WebCore/page/Page.cpp:
```diff
@@ -809,11 +809,13 @@ bool Page::findString(const String& target, FindOptions options, DidWrap* didWra
 
 #if ENABLE(IMAGE_ANALYSIS)
-void Page::analyzeImagesForFindInPage()
+void Page::setImageQueueEmptyHysteresisThenAnalyzeImagesForFindInPage(std::unique_ptr<PAL::HysteresisActivity>& timer)
```
`std::unique_ptr<PAL::HysteresisActivity>&&`

cdumez commented on Source/WebCore/page/Page.cpp:
```diff
@@ -809,11 +809,13 @@ bool Page::findString(const String& target, FindOptions options, DidWrap* didWra
     if (settings().imageAnalysisDuringFindInPageEnabled()) {
-        if (RefPtr mainDocument = m_mainFrame->document(); mainDocument && !m_imageAnalysisQueue)
+        if (RefPtr mainDocument = m_mainFrame->document(); mainDocument && !m_imageAnalysisQueue) {
+            imageAnalysisQueue().setImageQueueEmptyHysteresis(timer);
```
`WTFMove(timer)`

cdumez commented on Source/WebCore/page/Page.h:
```diff
@@ -975,7 +979,7 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
 
 #if ENABLE(IMAGE_ANALYSIS)
-    WEBCORE_EXPORT void analyzeImagesForFindInPage();
+    WEBCORE_EXPORT void setImageQueueEmptyHysteresisThenAnalyzeImagesForFindInPage(std::unique_ptr<PAL::HysteresisActivity>&);
```
`std::unique_ptr<PAL::HysteresisActivity>&&`

cdumez commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -293,11 +289,21 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
     RefPtr<WebPage> protectedWebPage = m_webPage;
+
+#if ENABLE(IMAGE_ANALYSIS)
+    auto timer = WTF::makeUnique<PAL::HysteresisActivity>([protectedWebPage, string, options, maxMatchCount] (PAL::HysteresisState state) {
```
`WTF::` is unnecessary.

cdumez commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -158,10 +159,19 @@ void ImageAnalysisQueue::resumeProcessing()
 
+void ImageAnalysisQueue::setImageQueueEmptyHysteresis(std::unique_ptr<PAL::HysteresisActivity>& timer)
+{
+    if (!m_imageQueueEmptyHysteresis)
```
Something looks fishy here. There is a single ImageAnalysisQueue per Page and the hysteresis timer activity has a lambda that captures the string being searched. So, in this scenario:
1. findString("foo")
2. findString("bar")

Step 1 will create the timer and search for "foo" in a one second timer. Then step 2 will come and won't update the timer. Therefore, after I've searched for "bar", then page will asynchronously override my search and search for "foo" again. This seems wrong.

Seems we should reset the hysteresis timer but It doesn't look like we do. Did I miss anything?

danae404 replied:
The timer only needs to be initialized and run the first time you do find-in-page on the webpage. If you load a webpage and do findString("foo"), at first only matches from the regular text on the page will appear, because the images on the page have not been analyzed yet. So to include the matches from the images, you need to analyze the images then perform findString("foo") again when the analysis completes. When you then do findString("bar"), the images on the page have already been analyzed from when you did findString("foo"). So there is no need to re-enqueue the images on the page and no need to initialize a timer that will redo findString("bar").

cdumez replied:
I don't think you understood my statement. What if the client searches for "foo", then for "bar" *before* your 1 second timer (which was scheduled for the "foo" search) fires?

I think it will do:
1. WebPage::findString("foo") from client
2. WebPage::findString("bar") from client
3. WebPage::findString("foo") from your timer scheduled at 1.

Seems wrong to use "foo" instead of "bar" at step 3 no?

danae404 replied:
Ah ok, I see now. Thank you, I will look into fixing that!

cdumez replied:
> There is no need to redo the image analysis, and so we don't need to worry about initializing a timer that will redo findString("bar").

I am not convinced this statement is correct. Pages are usually not static. Images can be dynamically added to the page at any time, which will cause `Page::didFinishLoadingImageForElement()` to get called. It will add the image to the queue and your timer WILL restart when the queue empties (again).

danae404 replied:
Good point. Thank you!

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
```diff
@@ -268,6 +269,20 @@
 static unsigned gDidProcessRequestCount = 0;
 
+@interface FindDelegate : NSObject<_WKFindDelegate> {
+    @public int _numCallsToDidFindMatches;
```
Though this is just testing code, it would be nice to clean this up using (more modern) ObjC idioms:

```
@interface FindDelegate : NSObject<_WKFindDelegate>
@property (nonatomic, readonly) int numberOfCallsToDidFindMatches;
@end

@implementation FindDelegate

 - (void)_webView:(WKWebView *)webView didFindMatches:(NSUInteger)matches forString:(NSString *)string withMatchIndex:(NSInteger)matchIndex
 {
     _numCallsToDidFindMatches++;
 }

 @end
```

(…and later…)

```
        return (gDidProcessRequestCount == 1 && [findDelegate numberOfCallsToDidFindMatches] == 2);
     }, 1, @"Timed out waiting for image analysis to start and for search for string to be performed twice.");
```

cdumez commented on Source/WebCore/page/ImageAnalysisQueue.h:
```diff
@@ -51,6 +55,10 @@ class ImageAnalysisQueue {
 
+    WEBCORE_EXPORT void setImageQueueEmptyHysteresis(std::unique_ptr<PAL::HysteresisActivity>&&);
+
+    bool analysisOfAllImagesOnPageHasStarted() { return m_analysisOfAllImagesOnPageHasStarted; }
```
Should be const:
`bool analysisOfAllImagesOnPageHasStarted() const { return m_analysisOfAllImagesOnPageHasStarted; }`

cdumez commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -293,11 +289,23 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
+
+#if ENABLE(IMAGE_ANALYSIS)
+    if (completionHandler) {
+        auto timer = makeUnique<PAL::HysteresisActivity>([protectedWebPage, string, options, maxMatchCount] (PAL::HysteresisState state) {
```
I'd feel better if we captured the WebPage as a WeakPtr, to be 100% sure we don't have a retain-cycle:
WebPage -> Page -> ImageAnalysisQueue -> back to the WebPage via the HystererisActivity lambda.

cdumez commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -293,11 +289,23 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
+            if (state == PAL::HysteresisState::Stopped)
+                protectedWebPage->findController().findString(string, options, maxMatchCount);
+        }, 1_s);
+        m_webPage->corePage()->setImageQueueEmptyHysteresisThenAnalyzeImagesForFindInPage(WTFMove(timer));
```
I'll defer to area experts here but I feel my this code might look better if it looked something like:
```
if (m_webPage->corePage()->needsImageAnalysisForFindInPage()) {
    m_webPage->corePage()->analyzeImagesForFindInPage([weakPage = WeakPtr { m_webPage.get() }, string, options, maxMatchCount] {
        if (weakPage)
            weakPage->findController().findString(string, options, maxMatchCount);
    });
}
```

The HysteresisActivity would be fully internal to the ImageAnalysisQueue.
`needsImageAnalysisForFindInPage()` would check something like:
`!analysisOfAllImagesOnPageHasStarted || hysteresisActivityIsStillRunning`

What do you think?

cdumez replied:
Wenson pointed out to me offline that the lambda may get called several times if images get added to the document while we're searching. However, if we're no longer displaying search results, we should definitely no longer call the lambda. I don't think this patch gets this quite right?

In any case, I think it'd be nice if the HysteresisActivity was internal complexity to the AnalysisQueue and here we'd just pass in a lambda that would get called when new images have been processed. And we should probably clear this lambda when we're done searching in page.

whsieh commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -293,11 +289,23 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
     RefPtr<WebPage> protectedWebPage = m_webPage;
+
+#if ENABLE(IMAGE_ANALYSIS)
+    if (completionHandler) {
```
Instead of basing this code off of the presence of `completionHandler`, it might be nicer to take a named `enum class` instead, like:

```
enum class TriggerImageAnalysis : bool { No, Yes };
```

whsieh commented on Source/WebCore/page/Page.cpp:
```diff
@@ -809,10 +809,11 @@ bool Page::findString(const String& target, FindOptions options, DidWrap* didWra
-        if (RefPtr mainDocument = m_mainFrame->document(); mainDocument && !m_imageAnalysisQueue)
+        imageAnalysisQueue().setImageQueueEmptyHysteresis(WTFMove(callback));
+        if (RefPtr mainDocument = m_mainFrame->document(); mainDocument && !imageAnalysisQueue().analysisOfAllImagesOnPageHasStarted())
             imageAnalysisQueue().enqueueAllImages(*mainDocument, { }, { });
```
I think it would be cleaner to rename this method to `enqueueAllImagesIfNeeded()`, and move the `m_analysisOfAllImagesOnPageHasStarted` check into `enqueueAllImagesIfNeeded()`. That way, each call site doesn't need to individually consult `analysisOfAllImagesOnPageHasStarted()`.

whsieh commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -158,10 +161,26 @@ void ImageAnalysisQueue::resumeProcessing()
     }
 }
 
+void ImageAnalysisQueue::setImageQueueEmptyHysteresis(Function<void()>&& callback)
```
Since this just takes a callback now, I'd rename this to `setDidBecomeEmptyCallback`.

whsieh commented on Source/WebCore/page/ImageAnalysisQueue.h:
```diff
@@ -51,6 +55,11 @@ class ImageAnalysisQueue {
+    WEBCORE_EXPORT void setImageQueueEmptyHysteresis(Function<void()>&& callback);
+    WEBCORE_EXPORT void endImageQueueEmptyHysteresis();
+
+    bool analysisOfAllImagesOnPageHasStarted() const { return m_analysisOfAllImagesOnPageHasStarted; }
```
(You can then remove this method, if you roll the check into `enqueueAllImages`).

whsieh commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -158,10 +161,26 @@ void ImageAnalysisQueue::resumeProcessing()
+    }, 1_s);
+}
+
+void ImageAnalysisQueue::endImageQueueEmptyHysteresis()
```
Similarly, this could just be `clearDidBecomeEmptyCallback()`.

whsieh reviewed (commented):
Looking much better!

whsieh commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -158,10 +164,26 @@ void ImageAnalysisQueue::resumeProcessing()
+
 void ImageAnalysisQueue::clear()
 {
     // FIXME: This should cancel pending requests in addition to emptying the task queue.
```
Let's reset `m_analysisOfAllImagesOnPageHasStarted` to `false` and `m_imageQueueEmptyHysteresis` to `nullptr` in `ImageAnalysisQueue::clear()`, as well.

whsieh commented on Source/WebCore/page/Page.cpp:
```diff
@@ -809,11 +809,12 @@ bool Page::findString(const String& target, FindOptions options, DidWrap* didWra
-        if (RefPtr mainDocument = m_mainFrame->document(); mainDocument && !m_imageAnalysisQueue)
-            imageAnalysisQueue().enqueueAllImages(*mainDocument, { }, { });
+        imageAnalysisQueue().setDidBecomeEmptyCallback(WTFMove(callback));
+        if (RefPtr mainDocument = m_mainFrame->document(); mainDocument)
```
Nit - just `if (RefPtr mainDocument = m_mainFrame->document())` here.

whsieh commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -233,16 +233,12 @@ void FindController::updateFindUIAfterPageScroll(bool found, const String& strin
 }
 
-void FindController::findString(const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, CompletionHandler<void(bool)>&& completionHandler)
+void FindController::findString(const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, TriggerImageAnalysis shouldTriggerImageAnalysis, CompletionHandler<void(bool)>&& completionHandler)
```
Nit - I would maybe call this enum `CanTriggerImageAnalysis` instead, since whether we end up doing so depends on whether the feature flag is set.

whsieh commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -291,13 +287,22 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
+        m_webPage->corePage()->setImageQueueEmptyHysteresisThenAnalyzeImagesForFindInPage([weakPage = WeakPtr { m_webPage }, string, options, maxMatchCount] {
+            if (weakPage)
+                weakPage->findController().findString(string, options, maxMatchCount, TriggerImageAnalysis::No); 
+        });
```
Nit - I still prefer @cdumez's earlier naming suggestion of:

```
m_webPage->corePage()->analyzeImagesForFindInPage([weakPage = WeakPtr { m_webPage.get() }, string, options, maxMatchCount] {
    if (weakPage)
        weakPage->findController().findString(string, options, maxMatchCount, TriggerImageAnalysis::No);
});
```

whsieh commented on Source/WebKit/WebProcess/WebPage/FindController.h:
```diff
@@ -51,6 +51,8 @@ class CallbackID;
 class PluginView;
 class WebPage;
 
+enum class TriggerImageAnalysis : bool { No, Yes };
```
Nit - maybe we should move this enum class into the `FindController` class scope? (Since it's a somewhat general name, but it has a very specific purpose here).

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
```diff
@@ -314,8 +331,8 @@ static void processRequestWithResults(id, SEL, VKImageAnalyzerRequest *, void (^
     TestWebKitAPI::Util::waitForConditionWithLogging([&] {
-        return gDidProcessRequestCount == 1;
-    }, 1, @"Timed out waiting for image analysis to start.");
+        return (gDidProcessRequestCount == 1 && [findDelegate numberOfCallsToDidFindMatches] == 2);
```
Nit - you can omit parentheses around the return value here.

danae404 committed:
Update find UI when ImageAnalysisQueue empties
https://bugs.webkit.org/show_bug.cgi?id=243071
rdar://97396153

Reviewed by Wenson Hsieh.

Test:
Updated FindInPageAPI.FindAPITextInImage to check that the first
Find-in-page operation on a webpage occurs twice.

This patch improves the internal setting ImageAnalysisForFindInPage.
Now the find ui is updated once the ImageAnalysisQueue is empty, so
matches from images show up shortly after other text matches without
the need for user interaction to update the ui. This is accomplished
by re-doing the most recent find-in-page when the image queue is newly
empty.

* Source/WebCore/PAL/pal/HysteresisActivity.h:

Added WTF_MAKE_FAST_ALLOCATED.

* Source/WebCore/page/ImageAnalysisQueue.cpp:
(WebCore::ImageAnalysisQueue::enqueueAllImages): Deleted.
(WebCore::ImageAnalysisQueue::enqueueAllImagesIfNeeded): Added.

Changed name of function and added m_analysisOfAllImagesOnPageHasStarted
Check.

(WebCore::ImageAnalysisQueue::resumeProcessing):

Now checks if the image queue is empty, and if so, triggers
m_imageQueueEmptyHysteresis.

(WebCore::ImageAnalysisQueue::setDidBecomeEmptyCallback): Added.

Initializes m_imageQueueEmptyHysteresis with given callback to be called
When the HysteresisActivity goes off.

(WebCore::ImageAnalysisQueue::clearDidBecomeEmptyCallback): Added.

(WebCore::ImageAnalysisQueue::clear):

Reset m_analysisOfAllImagesOnPageHasStarted and m_imageQueueEmptyHysteresis.

Sets m_imageQueueEmptyHysteresis to nullptr.

* Source/WebCore/page/ImageAnalysisQueue.h:

* Source/WebCore/page/Page.cpp:
(WebCore::Page::analyzeImagesForFindInPage):

If setting is enabled, calls ImageAnalysisQueue::setImageQueueEmptyHysteresis
with given callback, then enqueues all images on page
to be analyzed if they have not begun to be analyzed yet.

* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::findString):

Calls page::analyzeImagesForFindInPagee with a
callback function that calls FindController::findString with
Arguments configured to re-find the current string. Therefore there is a second
search for matches on the page and a UI update after the text from the images
has been added to the page.

findString has also been changed so that its completionHandler argument
Has a default value of { } and to take an enum shouldTriggerImageAnalysis.

(WebKit::FindController::hideFindUI):

Now also calls ImageAnalysisQueue::endImageQueueEmptyHysteresis.

* Source/WebKit/WebProcess/WebPage/FindController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp
(WebKit::WebPage::findString):

Now passes TriggerImageAnalysis::Yes to FindController::findString.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
(-[FindDelegate didFindMatches:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/image-with-text.html: Added.

Canonical link: https://commits.webkit.org/253244@main
